Readme documentation (#9)
ResultadosDigitais/bigtable-autoscaler-operator: README.md

[![CircleCI](https://circleci.com/gh/ResultadosDigitais/bigtable-autoscaler-operator.svg?style=svg)](https://circleci.com/gh/ResultadosDigitais/bigtable-autoscaler-operator)
[![GitHub release](https://img.shields.io/github/v/release/ResultadosDigitais/bigtable-autoscaler-operator.svg)](https://github.com/ResultadosDigitais/bigtable-autoscaler-operator/releases/latest)

# Bigtable Autoscaler Operator 
**Bigtable Autoscaler Operator** is a [Kubernetes Operator](https://kubernetes.io/docs/concepts/extend-kubernetes/operator/) to autoscale the number of nodes of a [Google Cloud Bigtable](https://cloud.google.com/bigtable) instance based on the CPU utilization.

## Overview
[Google Cloud Bigtable](https://cloud.google.com/bigtable) is designed to scale horizontally, meaning that the number of nodes of an instance can be increased to balance and reduce the average CPU utilization.

> For Bigtable applications dealing with high variances of workload, the procedure of managing the amount of nodes must be fully automated for a better resource cost.

Reviewer: ```suggestion
For Bigtable applications dealing with high variances of workload, automating the cluster scaling allow handling short load bursts while keeping costs as low as possible.
```

> This operator automates Bigtable scaling by reconciling the number of nodes and the CPU utilization with the `minNodes`, `maxNodes`, and `targetCPUUtilization` manifest specifications.

Reviewer: ```suggestion
This operator automates the scaling by balancing the number of nodes to keep the CPU utilization under the manifest specifications.
```

When the CPU utilization is above the target, the reconciler will increase the amount of nodes in steps linearly proportional to how above it is from the target.
For example, considering 100% of CPU utilization and only one node running, if the CPU target is 50%, it increases to 2 nodes, but if the CPU target is 25% it increases to 4 nodes.

The downscale also follows a linear rule, but it considers the `maxScaleDownNodes` specification which defines the maximum downscale step size in order to avoid aggressive downscale.
Furthermore, the downscale step is calculated using the amount of current nodes running and the CPU target. For example, if there are two nodes running and the CPU target is 50%, in order to downscale
occur the CPU utilization must go bellow 25%. This is important to avoid downscale that immediately causes upscale.

All scale operations are made respecting a reaction time window, which at time is not part of the manifest specification.

The image bellow shows how peaks above the CPU target of 50% are shortened by the automatic increase of nodes.
![Bigtable CPU utilization and nodes count](cpu_scaling.png "Autoscaling on CPU utilization.")

## Prerequisites

> 1. Enable APIs [Bigtable](https://cloud.google.com/bigtable/docs/access-control) and [Monitoring](https://cloud.google.com/monitoring/api/enable-api) on your GCP project.

Reviewer 2: ```suggestion
1. Enable [Bigtable](https://cloud.google.com/bigtable/docs/access-control) and [Monitoring](https://cloud.google.com/monitoring/api/enable-api) APIs on your GCP project.
```

    kubectl apply -f all-in-one.yml
    ```

## How to use this operator in a Kubernetes cluster
1. Use the service account from the [Prerequisites](#prerequisites) section to create the k8s secret
    ```sh
    kubectl create secret generic bigtable-autoscaler-service-account --from-file=service-account=./your_service_account.json
    ```

1. Apply a manifest based on the [sample file](./config/samples/bigtable_v1_bigtableautoscaler.yaml)
    ``` sh
    kubectl apply -f my_bigtableautoscaler.yaml
    ```
    
1. Check the autoscaler status
    ``` sh
    kubectl get bigtableautoscaler
    ```

## Development environment
These are the steps for setting up the development environment. 

This project is using go version 1.13 and other tools with its respective version, we don't guarantee that using other versions can perform successful builds.

1. Install [kubebuilder](https://book.kubebuilder.io/quick-start.html#installation) version 2.3.2. 
    1. Also make sure that you have its dependencies installed: [controller-gen](https://github.com/kubernetes-sigs/controller-tools/) version 0.5.0 and [kustomize](https://kubectl.docs.kubernetes.io/installation/kustomize/) version 3.10.0 

> 1. Follow [Option 1](#option-1) or [Option 2](#option-2) section.

Reviewer 2: ```suggestion
1. Follow [Option 1](#option-1) or [Option 2](#option-2) section.
1. Create your secret following [Secret Setup](#secret-setup) section.
```

Author: Actually, this secrete-setup section also says to you to create a role binding, but I think this not needed here. Right?

Reviewer 2: This are dev instructions right? We have to create the role binding at dev.

Author: Ops, yes, but I'm also referring this section in the correct step for each option.

Reviewer 2: whops, that's correct. Sorry!

- [Bigtable Autoscaler Operator](#bigtable-autoscaler-operator)
  * [Overview](#overview)
  * [Prerequisites](#prerequisites)
  * [How to install this operator in a Kubernetes cluster](#how-to-install-this-operator-in-a-kubernetes-cluster)
  * [How to use this operator in a Kubernetes cluster](#how-to-use-this-operator-in-a-kubernetes-cluster)
  * [Development environment](#development-environment)
    + [Option 1: Run with Tilt (recomended)](#option-1--run-with-tilt--recomended-)
    + [Option 2: Manual run](#option-2--manual-run)
  * [Secret setup](#secret-setup)
  * [Running tests](#running-tests)

## Overview
[Google Cloud Bigtable](https://cloud.google.com/bigtable) is designed to scale horizontally, meaning that the number of nodes of an instance can be increased to balance and reduce the average CPU utilization.
For Bigtable applications dealing with high variances of workload, the procedure of managing the amount of nodes must be fully automated for a better resource cost.
This operator automates Bigtable scaling by reconciling the number of nodes and the CPU utilization with the `minNodes`, `maxNodes`, and `targetCPUUtilization` manifest specifications.

> The reconciler behavior is to keep the CPU utilization of the instance bellow the target specification, it does that while respecting the minimum and maximum amount of nodes.

Reviewer: ```suggestion
The reconciler's responsibility is to keep the CPU utilization of the instance below the target specification respecting the minimum and maximum amount of nodes.
```
